### PR TITLE
Updated Observer Documentation

### LINKED ISSUE #1048: Improve documentation of threading with the observer pattern
### Description of the task

We currently don't do a good job of explaining the threading going on with our Observer design pattern for high-level communication between modules. We should add documentation for it as well as the `thread-safe-buffer` (probably)

### Acceptance criteria

- [ ] Be more explicit that in our usage of the `Observer` design pattern, `Obvservers` are what listen for and consume data, and `Subjects` create and publish the data, just so it's more immediately visible than in the middle of the linked article
- [ ] Explain how each Observer / module is running in it's own thread, and that each observer / subject template item runs in its own thread and has a callback called asynchronsly
- [ ] Possibly required for the above, explain the thread safe buffer a little bit

### PR DESCRIPTION
**Pull request by garethellis0**
### Description
For everyone reviewing, please let @jsn0716 @phelanm20 and @scveloso all review before reviewing. It's really important to get unbiased feedback from people who have no experience with this, as that's who the documentation is targeted at.

I've kept the first revision intentionally short, I'm relying on reviewers to point out where more detail / info is required!

### Resolved Issues
resolves #1048

### Review Checklist

**_It is the reviewers responsibility to also make sure every item here has been covered_**

- [x] **Function & Class comments**: All function definitions (usually in the `.h` file) should have a javadoc style comment at the start of them. For examples, see the functions defined in `thunderbots/software/geom`. Similarly, all classes should have an associated Javadoc comment explaining the purpose of the class.
- [x] **Remove all commented out code**
- [x] **Remove extra print statements**: for example, those just used for testing
- [x] **Resolve all TODO's**: All `TODO` (or similar) statements should either be completed or associated with a github issue
- [x] **Justify drops in code coverage**: If this PR results in a non-trivial drop in code coverage (a bot should post a coverage diagram as a comment), please justify why we can't test the code that's not covered.

### CODE REVIEW

File: docs/architecture-and-design.md [hunk 1]
```diff
@@ -171,12 +171,12 @@ Examples of the Visitor Pattern can be found with the following classes:
...
+The Observer Design Pattern is useful for letting components of a system "notify" each other when something happens. Read [https://www.geeksforgeeks.org/observer-pattern-set-1-introduction/] for a general introduction to the pattern.
 
-Read [https://www.geeksforgeeks.org/observer-pattern-set-1-introduction/] for more information.
-
-We use the Observer Design Pattern to connect all our top-level modules in the system as shown in the [Architecture Overview](#architecture-overview). The components can notify each other and provide data whenever it is available. For example, the [AI](#ai) will "observer" the [Backend](#backend) so that when the backend gets new information, it can notify the [AI](#ai) and the [AI](#ai) can choose to take action with the new data.
+Our implementation of this pattern consists of two classes, `Observer` and `Subject`. `Observer`'s can be registered with a `Subject`, after which new values will be sent from each `Subject` to all of it's registered `Observer`'s. Please see the headers of both classes for details. Note that a class can extend both `Observer` and `Subject`, thus receiving and sending out data. In this way we can "chain" multiple classes.
```
Comment (jsn0716) [inline docs/architecture-and-design.md]: `Observer`'s -> `Observer`s

**Review by jsn0716 — CHANGES_REQUESTED** [review]
I read through the observer documentation, the GeeksforGeeks link, and the headers for observer and subject, and I think the design pattern is overall clear to me. A quick specific example of this pattern might be good, as in the previous example with AI observing the Backend for new information.

File: docs/architecture-and-design.md (hunk 1 above)
Comment (garethellis0) [reply docs/architecture-and-design.md]: Changed and gave a quick example. @jonathanlew - just a note that the example will need to be updated when the "Sensor Fusion" layer exists.

**Comment by MathewMacDougall** [comment]
@phelanm20 @scveloso please review

**Comment by phelanm20** [comment]
Looks great! Very clear. For line 181, it would help the example if you were more specific. Specifying the subject 'ThreadedObserver' is receiving updates from, and giving an example operation it may then run would help. Pretty nit picky

**Comment by garethellis0** [comment]
Hmm, there's was a short example above, I re-positioned below the `ThreadedObserver` section, does that help make things more clear (please do feel free to say no if it's still at all unclear, better to fix this now then explain it 100x times later)?

**Comment by phelanm20** [comment]
Example looks great! Definitely helps clarify how we use the Observer pattern!

File: docs/architecture-and-design.md [hunk 2]
```diff
@@ -171,12 +171,15 @@ Examples of the Visitor Pattern can be found with the following classes:
...
 
-We use the Observer Design Pattern to connect all our top-level modules in the system as shown in the [Architecture Overview](#architecture-overview). The components can notify each other and provide data whenever it is available. For example, the [AI](#ai) will "observer" the [Backend](#backend) so that when the backend gets new information, it can notify the [AI](#ai) and the [AI](#ai) can choose to take action with the new data.
+### Threaded Observer
+In our system, we need to be able to do multiple things (receive camera data, run the AI, send commands to the robots) at the same time. In order to facilitate this, we extend the `Observer` to the `ThreadedObserver` class. The `ThreadedObserver` starts a thread with an infinite loop that waits for new data from `Subject` and performs some operation with it.
```
Comment (jonathanlew) [inline docs/architecture-and-design.md:179]: Can we discuss some pitfalls here? For example, observing multiple subjects at a time and potential data races. Also clarify `Subject`s don't run in their own threads (assuming that's the case)

**Review by jonathanlew — CHANGES_REQUESTED** [review]
Definitely clarifies our architecture a bit. Multithreading is tricky so we should be more proactive in warning about pitfalls.

**Review by scveloso — DISMISSED** [review]
Looks good to me, clear example of how to use it within our codebase

File: docs/architecture-and-design.md (hunk 2 above)
Comment (MathewMacDougall) [reply docs/architecture-and-design.md:179]: +1 for a brief overview of pitfalls, like locking variables handled by multiple observer callbacks. I think this gets close to being something that should be in-code documentation too, but I think very important to talk about it at a high-level here at least.

Comment (garethellis0) [reply docs/architecture-and-design.md:179]: Added a section with a warning and recommendation

**Review by jonathanlew — APPROVED** [review]
LGTM! :tada:

**Review by MathewMacDougall — APPROVED** [review]
Nice :book:

**Comment by garethellis0** [comment]
@jsn0716 - please re-review!

**Review by jsn0716 — APPROVED** [review]
LGTM! 👍

**Pull request merged**